evm/src/memory/memory_stark.rs

```diff
@@ -0,0 +1,551 @@
+                trace_rows[j][i] = val;
+            }
+        }
+        trace_rows
+    }
+
+    fn generate_memory(&self, trace_cols: &mut [Vec<F>]) {
+        let num_trace_rows = trace_cols[0].len();
+
+        let timestamp = &trace_cols[TIMESTAMP];
+        let is_read = &trace_cols[IS_READ];
+        let context = &trace_cols[ADDR_CONTEXT];
+        let segment = &trace_cols[ADDR_SEGMENT];
+        let virtuals = &trace_cols[ADDR_VIRTUAL];
+        let values: Vec<[F; 8]> = (0..num_trace_rows)
+            .map(|i| {
+                let arr: [F; 8] = (0..8)
+                    .map(|j| &trace_cols[value_limb(j)][i])
+                    .cloned()
+                    .collect_vec()
+                    .try_into()
+                    .unwrap();
+                arr
+            })
+            .collect();
+
+        let (
+            sorted_timestamp,
+            sorted_is_read,
+            sorted_context,
+            sorted_segment,
+            sorted_virtual,
+            sorted_values,
+        ) = sort_memory_ops(timestamp, is_read, context, segment, virtuals, &values);
+
+        let (context_first_change, segment_first_change, virtual_first_change) =
+            generate_first_change_flags(&sorted_context, &sorted_segment, &sorted_virtual);
+
+        let range_check_value = generate_range_check_value(
+            &sorted_context,
+            &sorted_segment,
+            &sorted_virtual,
+            &sorted_timestamp,
+            &context_first_change,
+            &segment_first_change,
+            &virtual_first_change,
+        );
+
+        trace_cols[SORTED_TIMESTAMP] = sorted_timestamp;
+        trace_cols[SORTED_IS_READ] = sorted_is_read;
+        trace_cols[SORTED_ADDR_CONTEXT] = sorted_context;
+        trace_cols[SORTED_ADDR_SEGMENT] = sorted_segment;
+        trace_cols[SORTED_ADDR_VIRTUAL] = sorted_virtual;
+        for i in 0..num_trace_rows {
+            for j in 0..8 {
+                trace_cols[sorted_value_limb(j)][i] = sorted_values[i][j];
+            }
+        }
+
+        trace_cols[CONTEXT_FIRST_CHANGE] = context_first_change;
+        trace_cols[SEGMENT_FIRST_CHANGE] = segment_first_change;
+        trace_cols[VIRTUAL_FIRST_CHANGE] = virtual_first_change;
+
+        trace_cols[RANGE_CHECK] = range_check_value;
+        trace_cols[COUNTER] = (0..trace_cols[0].len())
+            .map(|i| F::from_canonical_usize(i))
+            .collect();
+
+        let (permuted_inputs, permuted_table) =
+            permuted_cols(&trace_cols[RANGE_CHECK], &trace_cols[COUNTER]);
+        trace_cols[RANGE_CHECK_PERMUTED] = permuted_inputs;
+        trace_cols[COUNTER_PERMUTED] = permuted_table;
+    }
+
+    pub fn generate_trace(
+        &self,
+        memory_ops: Vec<(F, F, F, F, F, [F; 8])>,
+    ) -> Vec<PolynomialValues<F>> {
+        let mut timing = TimingTree::new("generate trace", log::Level::Debug);
+
+        // Generate the witness, except for permuted columns in the lookup argument.
```

Comment:
"except for permuted columns in the lookup argument" Is that up to date? Seems like these columns are generated.